### PR TITLE
Refactor task queue retrieval to be more robust and error prone

### PR DESCRIPTION
**Pull request by wonrax**
## Type

- [ ] Feature
- [ ] Bug fix
- [ ] CI/CD
- [ ] Documentation
- [ ] Other

## Description

Brief description of the changes made.

### CODE REVIEW

File: packages/mina-service/src/service/task-service.ts [hunk 1]
```diff
@@ -28,29 +42,52 @@ export class TaskService {
...
-          return true;
-        }
+      let backoff = true;
 
-        logger.debug('Task received:', task);
+      const task = await withCompoundTransaction(async (session) =>
+        ModelQueueTask.getInstance().acquireNextTaskInQueue(
+          session.proofService
+        )
+      );
```
Comment (SangTran-127) [inline packages/mina-service/src/service/task-service.ts:51]:
```suggestion
      const task = await withTransaction(async (session) =>
        ModelQueueTask.getInstance().acquireNextTaskInQueue(
          session
        )
      , "proofService");
```

File: packages/mina-service/src/service/task-service.ts [hunk 2]
```diff
@@ -28,29 +42,52 @@ export class TaskService {
...
+            await Proof.create(task, session);
+            return false;
+          } catch (error) {
+            logger.error(`Error processing task with ID ${task._id}: ${error}`);
+            return true;
+          } finally {
+            const processedTask = await ModelQueueTask.getInstance().findOne({
+              _id: task._id,
+            });
+
```
Comment (SangTran-127) [inline packages/mina-service/src/service/task-service.ts:67]: I think we need `session` here

File: packages/mina-service/src/service/task-service.ts [hunk 3]
```diff
@@ -28,29 +42,52 @@ export class TaskService {
...
-          await Proof.create(task, session);
-          return false;
-        } catch (error) {
-          logger.error(`Error processing task with ID ${task._id}: ${error}`);
-          return true;
-        }
-      });
+      if (task !== null) {
+        backoff = await withCompoundTransaction(async (session) => {
+          logger.debug('Task received:', task);
```
Comment (SangTran-127) [inline packages/mina-service/src/service/task-service.ts:55]: if only use transaction from proof db, you should use withTransaction(fn, "proofService") instead

File: packages/mina-service/src/service/task-service.ts (hunk 2 above)
Comment (wonrax) [reply packages/mina-service/src/service/task-service.ts:67]: fixed

File: packages/mina-service/src/service/task-service.ts (hunk 3 above)
Comment (wonrax) [reply packages/mina-service/src/service/task-service.ts:55]: prover needs a compound session, for example, to fetch document metadata:
https://github.com/orochi-network/zkDatabase/blob/a0d5fb1b226d4219024eaa77462c736786b8b637/packages/mina-service/src/domain/proof.ts#L41-L47

File: packages/storage/src/database/global/queue.ts [hunk 4]
```diff
@@ -48,65 +48,42 @@ export class ModelQueueTask extends ModelGeneral<WithoutId<TQueueRecord>> {
...
+    // - A task stuck at Proving status will block all other tasks from the
+    // same database. Maybe consider a timeout?
+    return this.collection.findOneAndUpdate(
+      {
+        status: EProofStatusDocument.Queued,
+        databaseName: {
+          $nin: await this.collection.distinct(
+            'databaseName',
+            {
               status: EProofStatusDocument.Proving,
```
Comment (SangTran-127) [inline packages/storage/src/database/global/queue.ts:73]: how about Proved?

Comment (SangTran-127) [reply packages/storage/src/database/global/queue.ts:73]: I think we should filter the Proved also

File: packages/mina-service/src/service/task-service.ts (hunk 3 above)
Comment (SangTran-127) [reply packages/mina-service/src/service/task-service.ts:56]: I agree, Closed

File: packages/storage/src/database/global/queue.ts (hunk 4 above)
Comment (wonrax) [reply packages/storage/src/database/global/queue.ts:73]: no, this check is to make sure that:
```
    // - Not allow parallel processing of tasks from the same database
```

that is, if a database has any `Proving` task, we exclude them from qualified next task to pop

**Review by SangTran-127 — APPROVED** [review]

**Pull request merged**